File: app/src/main/java/com/test/xyz/demo/presentation/weather/presenter/WeatherPresenterImpl.java
package com.test.xyz.demo.presentation.weather.presenter;

import com.test.xyz.demo.R;
import com.test.xyz.demo.domain.interactor.weather.CityValidationException;
import com.test.xyz.demo.domain.interactor.weather.UserNameValidationException;
import com.test.xyz.demo.domain.interactor.weather.WeatherInteractor;
import com.test.xyz.demo.domain.model.weather.WeatherSummaryInfo;
import com.test.xyz.demo.presentation.common.DisposableManager;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;

public class WeatherPresenterImpl implements WeatherPresenter {
    private final WeatherView mainView;
    private final WeatherInteractor weatherInteractor;
    private final WeatherDataFormatter weatherDataFormatter;
    private final WeatherDegreeConverterProxy weatherDegreeConverterProxy;
    private final DisposableManager disposableManager;

    @Inject
    public WeatherPresenterImpl(WeatherView mainView, WeatherInteractor weatherInteractor,
                                WeatherDataFormatter weatherDataFormatter,
                                WeatherDegreeConverterProxy weatherDegreeConverterProxy) {

        this.mainView = mainView;
        this.weatherInteractor = weatherInteractor;
        this.weatherDataFormatter = weatherDataFormatter;
        this.weatherDegreeConverterProxy = weatherDegreeConverterProxy;
        this.disposableManager = new DisposableManager();
    }

    @Override
    public void requestWeatherInformation() {
        mainView.showBusyIndicator();

        Disposable disposable = weatherInteractor.getWeatherInformation(mainView.getUserNameText(), mainView.getCityText())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableObserver<WeatherSummaryInfo>() {
                    @Override
                    public void onNext(WeatherSummaryInfo weatherSummaryInfo) {
                        mainView.hideBusyIndicator();

                        weatherSummaryInfo.setTemperature(
                                weatherDegreeConverterProxy.convertFahrenheitToCelsius(weatherSummaryInfo.temperature())
                        );

                        mainView.showResult(weatherDataFormatter.format(weatherSummaryInfo));
                    }

                    @Override
                    public void onError(Throwable e) {
                        mainView.hideBusyIndicator();

                        if (e instanceof UserNameValidationException) {
                            mainView.showUserNameError(R.string.username_validation_error_message);
                            return;
                        }

                        if (e instanceof CityValidationException) {
                            mainView.showCityNameError(R.string.cityname_validation_error_message);
                            return;
                        }

                        mainView.showGenericError(R.string.weather_error);
                    }

                    @Override
                    public void onComplete() {
                    }
                });

        disposableManager.add(disposable);
    }

    @Override
    public void onStop() {
        disposableManager.dispose();
    }
}

File: app/src/main/kotlin/com/test/xyz/demo/presentation/weather/presenter/WeatherPresenterImpl.kt
package com.test.xyz.demo.presentation.weather.presenter

import com.test.xyz.demo.R
import com.test.xyz.demo.domain.interactor.weather.CityValidationException
import com.test.xyz.demo.domain.interactor.weather.UserNameValidationException
import com.test.xyz.demo.domain.interactor.weather.WeatherInteractor
import com.test.xyz.demo.domain.model.weather.WeatherSummaryInfo
import com.test.xyz.demo.presentation.common.DisposableManager
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableObserver
import javax.inject.Inject

class WeatherPresenterImpl @Inject
constructor(private val mainView: WeatherView, private val weatherInteractor: WeatherInteractor,
            private val weatherDataFormatter: WeatherDataFormatter,
            private val weatherDegreeConverterProxy: WeatherDegreeConverterProxy) : WeatherPresenter {

    private val disposableManager: DisposableManager

    init {
        this.disposableManager = DisposableManager()
    }

    override fun requestWeatherInformation() {
        mainView.showBusyIndicator()

        val disposable = weatherInteractor.getWeatherInformation(mainView.getUserNameText(), mainView.getCityText())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableObserver<WeatherSummaryInfo>() {
                    override fun onNext(weatherSummaryInfo: WeatherSummaryInfo) {
                        mainView.hideBusyIndicator()

                        weatherSummaryInfo.setTemperature(
                                weatherDegreeConverterProxy.convertFahrenheitToCelsius(weatherSummaryInfo.temperature().toFloat())
                        )

                        mainView.showResult(weatherDataFormatter.format(weatherSummaryInfo))
                    }

                    override fun onError(e: Throwable) {
                        mainView.hideBusyIndicator()

                        if (e is UserNameValidationException) {
                            mainView.showUserNameError(R.string.username_validation_error_message)
                            return
                        }

                        if (e is CityValidationException) {
                            mainView.showCityNameError(R.string.cityname_validation_error_message)
                            return
                        }

                        mainView.showGenericError(R.string.weather_error)
                    }

                    override fun onComplete() {}
                })

        disposableManager.add(disposable)
    }

    override fun onStop() {
        disposableManager.dispose()
    }
}
